maketeams3.py

```diff
@@ -258,60 +313,219 @@ def testSwap(teama, playera, teamb, playerb, board):
     for player in players:
         player.setReqMet()
 
-    #WIP for upload format for heltour
+    return {'teams': teams,
+            'players': players,
+            'alternates': alternates,
+            'team_rating_bounds': team_rating_bounds,
+            'alt_rating_bounds': alt_rating_bounds,
+            'alts_split': alts_split}
+
+
+# Reduce variance functions
+
+def intersection(lst1, lst2):
+    return set(lst1).intersection(set(lst2))
+
+
+# Does this swap have a neutral effect on happiness
+def is_neutral_swap(swap):
+    def count_on_team(attr, player, team):
+        n = len(intersection(getattr(player, attr), team.boards))
+        n += len([p for p in team.boards if player in getattr(p, attr)])
+        return n
+
+    count_friends_on_team = partial(count_on_team, 'friends')
+    count_avoids_on_team = partial(count_on_team, 'avoid')
+
+    pa, pb = swap
+    if count_friends_on_team(pa, pa.team) != count_friends_on_team(pa, pb.team)\
+            or count_friends_on_team(pb, pb.team) != count_friends_on_team(pb, pa.team)\
+            or count_avoids_on_team(pa, pa.team) != count_avoids_on_team(pa, pb.team)\
+            or count_avoids_on_team(pb, pb.team) != count_avoids_on_team(pb, pa.team):
+        return False
+
+    return True
+
+
+def get_swaps(teams):
+    num_boards = len(teams[0].boards)
+    boards = [[team.boards[i] for team in teams] for i in range(num_boards)]
+    swaps = [[swap for swap in combinations(board, 2) if is_neutral_swap(swap)] for board in boards]
+    return [swap for board_swaps in swaps for swap in board_swaps]
```

Comment:
This just flattens the list right?  I find the syntax and variables names to be unnecessarily confusing in this case. Maybe this is some well known python pattern that I'm not aware of, but I still find it confusing.

Reply:
YEah, I'm not a python guy really.  This is what I've been told told is the most "pythonic" and efficient way of flattening lists. I agree though. I have no idea how this works and dislike it tremendously. I will replace it with a function called flatten.

The alternative semi-pythonic way is `sum(list, [])` which is almost as bad.

Reply:
function called flatten clears it up for me and is what I would prefer. The implementation of that function can be any of the ones we've suggested. It's named and I can just believe it. :)